arangod/Cluster/Maintenance.cpp

```diff
@@ -0,0 +1,814 @@
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Kaveh Vahedipour
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Basics/StringUtils.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Logger/Logger.h"
+#include "Cluster/FollowerInfo.h"
+#include "Cluster/Maintenance.h"
+#include "Utils/DatabaseGuard.h"
+#include "VocBase/LogicalCollection.h"
+#include "VocBase/Methods/Databases.h"
+
+#include <velocypack/Compare.h>
+#include <velocypack/Iterator.h>
+#include <velocypack/Slice.h>
+#include <velocypack/velocypack-aliases.h>
+
+#include <algorithm>
+
+using namespace arangodb;
+using namespace arangodb::basics;
+using namespace arangodb::maintenance;
+using namespace arangodb::methods;
+
+static std::vector<std::string> cmp {
+  "journalSize", "waitForSync", "doCompact", "indexBuckets"};
+static std::string const CURRENT_COLLECTIONS("Current/Collections/");
+static std::string const CURRENT_DATABASES("Current/Databases/");
+static std::string const ERROR_MESSAGE("errorMessage");
+static std::string const ERROR_NUM("errorNum");
+static std::string const ERROR("error");
+static std::string const PLAN_ID("planId");
+static std::string const PRIMARY("primary");
+static std::string const SERVERS("servers");
+static std::string const SELECTIVITY_ESTIMATE("selectivityEstimate");
+static std::string const COLLECTIONS("Collections");
+static std::string const DB ("/_db/");
+static std::string const FOLLOWER_ID("followerId");
+static VPackValue const VP_DELETE("delete");
+static VPackValue const VP_SET("set");
+static std::string const OP("op");
+static std::string const UNDERSCORE("_");
+
+std::shared_ptr<VPackBuilder> createProps(VPackSlice const& s) {
+  auto builder = std::make_shared<VPackBuilder>();
+  TRI_ASSERT(s.isObject());
+  { VPackObjectBuilder b(builder.get());
+    for (auto const& attr : VPackObjectIterator(s)) {
+      std::string const key = attr.key.copyString();
+      if (key == ID || key == NAME) {
+        continue;
+      }
+      builder->add(key, attr.value);
+    }}
+  return builder;
+}
+
+
+std::shared_ptr<VPackBuilder> compareRelevantProps (
+  VPackSlice const& first, VPackSlice const& second) {
+  auto result = std::make_shared<VPackBuilder>();
+  { VPackObjectBuilder b(result.get());
+    for (auto property : cmp) {
+      auto const& planned = first.get(property);
+      if (planned != second.get(property)) { // Register any change
+        result->add(property,planned);
+      }
+    }
+  }
+  return result;
+}
+
+
+VPackBuilder compareIndexes(
```

Comment:
@jsteemann and @kvahed:  Looking for a second opinion here.  Return of a VPackBuilder object appears really expensive to me.  My understanding is that C++ will create a temporary VPackBuilder, copy the contents of local "builder" into the temporary, then copy the temporary object into calling function's VPackBuilder object.  Seems like a ton of overhead.  I know this is not likely on a critical path, but wondering if we should have caller pass in VPackBuilder reference just because we like cleaner code.  Your thoughts?

Reply:
shouldn't this use RVO ? Also I think a builder does have a move constructor, so it should move the underlying buffer if possible.

Reply:
I also think that usually (unless there is a good reason against it), it will use return value optimization.